### ISSUE TITLE
JitGenericAttribute doesn't work if arguments are reference types

### ISSUE THREAD

**Issue by ashmind** [ashmind/SharpLab#99]

**Comment by JosephTremoulet** [comment]
I'd guess this has to do with "generic sharing" - for reference types, we generate a single method body of jitted machine code and use that for all of them (it has extra indirections for things that depend on the specific reference type, like if the code has `new T() `). There's a special placeholder type  called `System. Canon` that represents "some unspecified reference type" which is what the VM tells the JIT is the type argument that the generic is being instantiated with in these cases. So I wouldn't be surprised if you're referencing a specific reference type in some api call where you need to be substituting `System. Canon` instead. But I'm not familiar enough with these interfaces to know which might be having the issue or how to reference `System. Canon` in it... 

(all of this is just a guess based on the statement that reference types are what's failing, btw)

**Comment by ashmind** [comment]
@JosephTremoulet Thanks! I actually considered that and tried to use `MakeGenericType` with `System.__Canon`, but I wasn't able to get it to work either. My current guess is that clrmd resolves shared generics incorrectly, but sadly that will require significant time to research/debug.

**Comment by ashmind** [comment]
The problem here is that `HotSize` is `0` btw (no cold either). I saw same issue with some other generics (inherited/non-open) where the same method was duplicated on a type -- once without proper `HotColdInfo` and once with it. I was able to resolve it by finding a correct duplicate. Might be related, but I can't apply the same solution here.

**Comment by JosephTremoulet** [comment]
Maybe you could try using the approach outlined [here](https://github.com/Microsoft/clrmd/blob/master/Documentation/MachineCode.md) and see if it gets better results than using `HotColdInfo`.

**Comment by ashmind** [comment]
This works in .NET Core branch, so closing as "cannot reproduce".

**Comment by canton7** [comment]
Just saw this message pop up, and noticed this issue was "cannot reproduce", so sharing:

```cs
using System;
using SharpLab.Runtime;
public class C {
    [JitGeneric(typeof(int)), JitGeneric(typeof(string))]
    public string M<T>() {
        if (typeof(T) == typeof(int))
            return "int";
        if (typeof(T) == typeof(string))
            return "string";
        return "neither";
    }
}
```

Compile as x64 (or Default/master), Release. [Link](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACAZQAtsoAHAGW2ADoASgFcAdhgCW+GAG4axAMwMyDAMIMA3jQbaGAbQBS4jAHEYImFHFgAFBgCeHGBABm18WICUHtA0MmzFla2Dk6uxOR0XgC6WjoKShEMALIAPAAqAHzWHhqxOjrizgzBji7WaTkAvJUM9qWu7hheefmtxADsDABEjV2y1K0FRSWh5VU1daPhkR4tg9od3dN9c3GdXeZGLBYrAzoAvjT7QA==).

**Comment by mstefarov** [comment]
[Another reproducer here.](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACY8gOgCUBXAOwwEt8YLAMIR8AB14AbGFADKMgG68wMXAG4a9BrIAW2KGIAy2YO258BG6poDMTckiakGI8fpgARAJ5ds+ZUJ4fFwA5gwA3jQM0QwA2gBSvBgA4jBcMsoAFBheYjAQAGaZvDwAlKVoDIkpaRlg2bn5Rcx05QC6UTHEdsyOACoMAKq4MINcuNgFMACCuAA8fQB8maWd0ZHUMVsMvAUMDXmFmX2lDAC8Zww5h83kraub2+trTzEK+gz4XrIYUOcMAFlsABrGA/KAlEIrKyvbbEADsQ3Gk0EszmLUqS0ysD2X3BpRhrwAvjBJCMdnsDk1jqcLldGkcShhyhtYVsmZ8vABJHj/IGggByHHwwBk0JesIRSImUxYaKZmOWOM5POZhKeJLJMFZbOiUoAJjACtgOJIMOrtkSXlbHtEXgkkql0hD6tdqUzypVqk66lSji12i9uvZ+kMRkI9KEYH1Ggtlg8tjqnrt9m6jidzpc07d7hKYkm2e8/njfnyQWDfpDxba2VKaSIuAoZBhhJGQtHGpkS1BKtmaQS89FNeSU36ihm6X2PaUC7COV9VWXBcLRVBq7q9Yj6xBG83W9gozG8l3uTxewzx+ULVth9rB3DEYbjabzfebTeXvaATAMDoIPquXESRpgwStgA4DAYEyb9f3/QCxEkAB5MQ+B3XAWAFCAeUkEpIVKNoGBeMQIXeSCQ3sRh+QrCFQhWc5FgYAAiYAIAgRiLS/H8/wAoCQLAiCoJg7j4KQlDeDQjCsK4HCuDwgiiJI7AyN6HZeSooURTFWkGPIKwiSAA===)

```cs
using System;
using System.Runtime.CompilerServices;
using SharpLab.Runtime;

public static class CompareDynamicCasting {
    [JitGeneric(typeof(int)), JitGeneric(typeof(string))]
    public static T UseUnsafeAs<T>()
    {
        if (typeof(T) == typeof(string))
        {
            var myStr = MakeString();
            return Unsafe.As<string, T>(ref myStr);
        }else if (typeof(T) == typeof(int)){
            int myInt = MakeNumber();
            return Unsafe.As<int, T>(ref myInt);
        }else{
            return default;
        }
    }
    
    [JitGeneric(typeof(int)), JitGeneric(typeof(string))]
    public static T UseChangeType<T>()
    {
        if (typeof(T) == typeof(string))
        {
            var myStr = MakeString();
            return (T)Convert.ChangeType(myStr, typeof(T));
        }else if (typeof(T) == typeof(int)){
            int myInt = MakeNumber();
            return (T)Convert.ChangeType(myInt, typeof(T));
        }else{
            return default;
        }
    }
    
    [MethodImplAttribute(MethodImplOptions.NoInlining)] 
    private static string MakeString() => "boo";
    
    [MethodImplAttribute(MethodImplOptions.NoInlining)] 
    private static int MakeNumber() => 1;
}
```

Viewing JIT Asm works for `int` but not for `string`.

**Comment by ashmind** [comment]
Thanks! I'll revisit it when I have a moment.

**Comment by kkokosa** [comment]
Yep, [simple case](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACAZQAtsoAHAGW2ADoASgFcAdhgCW+GAG4aNYgGYGZBgGEGAbwC+8pbgxQhYDM007aSlcXIB2GhpoMnDANoApcRgDiMETCjiYAAUGACeHDAQAGZBqgCUcWgMHt6+/oEh4ZExTAkAuo7OigzAEBAANgwAkrgAatjlQjAAKlkAPM0AfEFxDAC8nQxhEdFBzXF8NfWNLVmy1OZAA==):

```
using System;
using SharpLab.Runtime;

public class C {}
public struct S {}
public class Program
{
    [JitGeneric(typeof(C)), JitGeneric(typeof(S))]
    public bool IsValueType<T>() => typeof(T).IsValueType;
}
```

**Comment by SpiritBob** [comment]
Is there any way we can view JIT compiled code on our own? That way we'll at least be able to see what is actually being generated when we define a generic that accepts reference typed arguments.

**Comment by ashmind** [comment]
Just FYI, asked for some more clarification in https://github.com/dotnet/runtime/discussions/56274.

**Comment by rootflood** [comment]
i use MethodImplOptions.AggressiveInlining flag for this case [sharplab](https://sharplab.io/#v2:C4LghgzgtgPgAgJgIwFgBQcAMACAygCzACcAHAGTACMA6AJQFcA7YASygFMBudLbOJOk1YdqAYQD2UEiwA27IrnkA3FgGN2Ebmh4BmPgmyjsAb3TZz2Mxf4A2PkhxQkACn44AHgEoAvAD4A4uzAAGpgMvTsLl5aFpZosQDaAFIswIGM8mrOwACeJOziAGauDp6eALpW5gkAskH44gAmAJJSMs51wA0tbQDyJKzijBDUAIIA5uNEGhAsSuzNjDIsjCvj2DDYnd2tJDL9g8Njk9MQs/MHbCwAXmCHFbGPFlV8erb2OIEhYRFIADzAXzZbBKTzoF6meJPPgAdhB2AgAHdUqp8C9YpDoViPgjsH4EQAadHY8ziSgAK3YqmA2HEeN8tOoABVxLhgEQ1s4wVCngBfGIWXkvcE8+x2NzYKAIEoeHwBIKhcLsBB/NxA6IixIpNLsDIc1TZPIFYpuMqVUW1epNXbtbbWvoDFhDEYTKYzOYLJYrNYbLZWnp7S7O45us4ey5QG53J2MB4kuIvOBvJDihzYL6KiIqwHA0EQ4nmJTEbDAEh4ktGorZTwCp4sYqlvHeCv5Kum7nx2HYACqwzAhXYYwgAIJauc00KIJrBew7BkEHYM9icDhjXYhTA9BkwFr2CFovQvKAA==)

**Comment by timcassell** [comment]
I'm also seeing this issue with a struct containing a reference type.

```cs
[JitGeneric(typeof(KeyValuePair<int, object>))]
```
